lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -0,0 +1,833 @@
+    // constant exit conditions.
+    for (BasicBlock *ClonedBB : CurIteration.Body)
+      SimplifyInstructionsInBlock_NoDelete(ClonedBB, NULL);
+
+    // Check exit condition to see if we fully unrolled the loop
+    if (BranchInst *BI = dyn_cast<BranchInst>(CurIteration.Latch->getTerminator())) {
+      bool Cond = false;
+      if (!GetConstantI1(BI->getCondition(), &Cond))
+        break;
+
+      if (!Cond && BI->getSuccessor(0) == CurIteration.Header) {
+        Succeeded = true;
+        break;
+      }
+      else if (Cond && BI->getSuccessor(1) == CurIteration.Header) {
+        Succeeded = true;
+        break;
+      }
+    }
+  }
+
+  if (Succeeded) {
+    auto &FirstIteration = Clones.front();
+    // Go through the predecessors of the old header and
+    // make them branch to the new header.
+    SmallVector<BasicBlock *, 8> Preds(pred_begin(Header), pred_end(Header));
+    for (BasicBlock *PredBB : Preds) {
+      if (L->contains(PredBB))
+        continue;
+      BranchInst *BI = cast<BranchInst>(PredBB->getTerminator());
+      for (unsigned i = 0, NumSucc = BI->getNumSuccessors(); i < NumSucc; i++) {
+        if (BI->getSuccessor(i) == Header) {
+          BI->setSuccessor(i, FirstIteration.Header);
+        }
+      }
+    }
+
+    Loop *OuterL = L->getParentLoop();
+
+    if (OuterL) {
+      // Core body blocks need to be added to outer loop
+      for (size_t i = 0; i < Clones.size(); i++) {
+        auto &Iteration = Clones[i];
+        for (BasicBlock *BB : Iteration.Body) {
+          if (!Iteration.Extended.count(BB)) {
+            OuterL->addBasicBlockToLoop(BB, *LI);
+          }
+        }
+      }
+
+      // Our newly created exit blocks may need to be added to outer loop
+      for (BasicBlock *BB : FakeExits) {
+        if (HasSuccessorsInLoop(BB, OuterL))
+          OuterL->addBasicBlockToLoop(BB, *LI);
+      }
+
+      // Cloned exit blocks may need to be added to outer loop
+      for (size_t i = 0; i < Clones.size(); i++) {
+        auto &Iteration = Clones[i];
+        for (BasicBlock *BB : Iteration.Extended) {
+          if (HasSuccessorsInLoop(BB, OuterL))
+            OuterL->addBasicBlockToLoop(BB, *LI);
+        }
+      }
+    }
+
+    // Remove the original blocks that we've cloned from all loops.
+    for (BasicBlock *BB : ToBeCloned) {
+      LI->removeBlock(BB);
+    }
+
+    LPM.deleteLoopFromQueue(L);
+
+    // Remove dead blocks.
+    for (BasicBlock *BB : ToBeCloned)
+      DetachFromSuccessors(BB);
+    for (BasicBlock *BB : ToBeCloned)
+      BB->dropAllReferences();
+    for (BasicBlock *BB : ToBeCloned)
+      BB->eraseFromParent();
+
+    if (OuterL) {
+      // This process may have created multiple back edges for the
+      // parent loop. Simplify to keep it well-formed.
+      simplifyLoop(OuterL, DT, LI, this, nullptr, nullptr, AC);
+    }
+
+    return true;
+  }
+
+  // If we were unsuccessful in unrolling the loop
+  else {
```

Comment:
Is it legal to continue if we don't succeed unrolling an explicit [unroll] attribute or should we error out?